### DISCUSSION TITLE
Give parser function visibility of choices descriptions

### DISCUSSION THREAD

**Discussion by wsande** [python-cmd2/cmd2#1159] [Q&A]
Using an exampple similar to my previous question...

Let's say I have a 'show fruit' command with a completer that produces this when I tab:
```
ID
=========
1     apple (Fuji)
2     orange (Mandarin)
3     banana (Chiquita)
all   (show all fruit)
```

When I complete the command and the parser function is called, it gets the ID from the choice provider.
How can I also make the choice descr available to the parser function?

See code below, with my question in the `'show fruit'` function.

```
import cmd2
from typing import  List

def fruit_completer(self) -> List[cmd2.CompletionItem]:
    items = {1:'apple (Fuji)', 2:'orange(Mandarin)', 3:'banana (Chiquita)', 'all':'(show all fruit)' }
    return [cmd2.CompletionItem(item_id, description) for item_id, description in items.items()]

show_parser = cmd2.Cmd2ArgumentParser()
show_subparsers = show_parser.add_subparsers(title='subcommands', help='subcommand help')
parser_show_fruit = show_subparsers.add_parser('fruit', help='show fruit help')
fruit_arg = parser_show_fruit.add_argument('fruit',  choices_provider=fruit_completer, metavar='ID',  descriptive_header='',  help="enter a fruit ID or 'all'")

class SAMPLE_CLI(cmd2.Cmd):
    def __init__(self):
        super().__init__()

    def show_fruit(self, args):
        self.poutput("You chose fruit #{}".format(args.fruit))
        # -----
        # how do I access the arg choice descr string  here?  e.g. 'apple (Fuji)'
        # -----

    parser_show_fruit.set_defaults(func=show_fruit)

    @cmd2.with_argparser(show_parser)
    def do_show(self, args):
        func = getattr(args, 'func', None)
        if func is not None:
            func(self, args)
        else:
            self.do_help('show')

if __name__ == '__main__':
    import sys
    app = SAMPLE_CLI()
    sys.exit(app.cmdloop())
```

**Comment by kmvanbrunt** [comment]
`CompletionItem` data is only used during tab completion to display suggestions and populate the command line. What's passed to your parser is coming from the command line after you press Enter.

Therefore, you'll need to move the `items` dictionary out of `fruit_completer()` so it can also be accessed from `show_fruit()`.
Then just use the passed in fruit ID to resolve its description.

**Reply by wsande** [reply]
Thanks for that! One thing I realized is that if the user just types '`show fruit`' instead of using tab completion, the completer never runs (of course) and `items` never gets populated.  So even though I moved it out of `fruit_completer()`, it was empty.

I fixed this by explicitly calling `fruit_completer()` at the start of the `show_fruit()` function, so `items` is populated whether or not they used tab-completion.

**Comment by kmvanbrunt** [comment]
Why not just populate the dictionary when you declare it?

```Python
fruit_choices = {1:'apple (Fuji)', 2:'orange(Mandarin)', 3:'banana (Chiquita)', 'all':'(show all fruit)'}

def fruit_completer(self) -> List[cmd2.CompletionItem]:
    return [cmd2.CompletionItem(item_id, description) for item_id, description in fruit_choices.items()]
```

**Reply by wsande** [reply]
In my real CLI (not this simplified sample), the choices are dynamic, based on context and the current state of the system.
